synthesis/synthesis.go

```diff
@@ -0,0 +1,273 @@
+	codonInsert := `INSERT INTO codon(codon, aa) VALUES (?, ?)`
+	for _, aminoAcid := range codontable.AminoAcids {
+		for _, codon := range aminoAcid.Codons {
+			db.MustExec(codonInsert, codon.Triplet, aminoAcid.Letter)
+			weightTable[codon.Triplet] = codon.Weight
+
+			codonBias := strings.Count(codon.Triplet, "G") + strings.Count(codon.Triplet, "C")
+			for _, toCodon := range aminoAcid.Codons {
+				if codon.Triplet != toCodon.Triplet {
+					toCodonBias := strings.Count(toCodon.Triplet, "G") + strings.Count(toCodon.Triplet, "C")
+					switch {
+					case codonBias == toCodonBias:
+						db.MustExec(`INSERT INTO codonbias(fromcodon, tocodon, gcbias) VALUES (?, ?, ?)`, codon.Triplet, toCodon.Triplet, "NA")
+					case codonBias > toCodonBias:
+						db.MustExec(`INSERT INTO codonbias(fromcodon, tocodon, gcbias) VALUES (?, ?, ?)`, codon.Triplet, toCodon.Triplet, "AT")
+					case codonBias < toCodonBias:
+						db.MustExec(`INSERT INTO codonbias(fromcodon, tocodon, gcbias) VALUES (?, ?, ?)`, codon.Triplet, toCodon.Triplet, "GC")
+					}
+				}
+			}
+		}
+	}
+
+	// Insert seq and history
+	pos := 0
+	for i := 0; i < len(sequence); i = i + 3 {
+		codon := sequence[i : i+3]
+		db.MustExec(`INSERT INTO seq(pos) VALUES (?)`, pos)
+		db.MustExec(`INSERT INTO history(pos, codon, step) VALUES (?, ?, 0)`, pos, codon)
+		db.MustExec(`INSERT INTO weights(pos, codon, weight) VALUES (?,?,?)`, pos, codon, weightTable[codon])
+		pos++
+	}
+
+	var err error
+	// For a maximum of 100 iterations, see if we can do better. Usually sequences will be solved within 1-3 rounds,
+	// so 100 just effectively acts as the max cap for iterations. Once you get to 100, you pretty much know that
+	// we cannot fix the sequence.
+	for i := 1; i < FixIterations; i++ {
+		suggestions := findProblems(sequence, problematicSequenceFuncs)
+		// If there are no suggestions, break the iteration!
+		if len(suggestions) == 0 {
+			return sequence, nil
+		}
+		for _, suggestion := range suggestions { // if you want to add overlaps, add suggestionIndex
+			// First, let's insert the suggestions that we found using our problematicSequenceFuncs
+			_, err = db.Exec(`INSERT INTO suggestedfix(step, start, end, gcbias, quantityfixes, suggestiontype) VALUES (?, ?, ?, ?, ?, ?)`, i, suggestion.Start, suggestion.End, suggestion.Bias, suggestion.QuantityFixes, suggestion.SuggestionType)
+			if err != nil {
+				return sequence, err
+			}
+		}
+
+		// The following statements are the magic sauce that makes this all worthwhile.
+		// Parameters: step, gcbias, start, end, quantityfix
+		sqlFix1 := `INSERT INTO history
+		            (codon,
+		             pos,
+		             step)
+		SELECT t.codon,
+		       t.pos,
+		       ? AS step
+		FROM   (SELECT cb.tocodon AS codon,
+		               s.pos      AS pos
+		        FROM   seq AS s
+		               JOIN history AS h
+		                 ON h.pos = s.pos
+		               JOIN weights AS w
+		                 ON w.pos = s.pos
```

Comment:
> Using this line this will actually make all positions and codons with the same weight and you will not select the codons that have the highest weights and aren't the last ones used.

That line doesn't actually do any writing, so I can't really see how it could make all the positions / codons the same weight. What it should be doing is selecting all weights at a given position, which I do believe it does correctly

> Removing the pos field in the weight table, because the weights are dependent of the codon, but independent of position (you should have the same weight, for the same codon, even in different positions)

This was a specific design decision and implementation for codon harmonization purposes. I don't believe we should change this, and if it is necessary to push the PR through, we should go add codon harmonization now.